In [8]:
import xmldict
import xml.etree.ElementTree as ET
import re
from loguru import logger
import sys
from collections import namedtuple

In [9]:
Expression = namedtuple('Expression', ('prop', 'condition_type', 'expression'))

In [10]:
tree = ET.parse('../data_samples/export_22.07.2020 14_03_16.xml')

In [11]:
root = tree.getroot()

In [12]:
forms = list(root.iter('forms'))

In [13]:
forms_exprs_props = {}

property_expression_pattern = re.compile(r'(?<=<key>)(.*?)(?=</key>\s*<value(.*?)>)')
expression_pattern = re.compile(r'(\w+)=\"#{(.*?)}\"')

for form in forms:
    props_exprs = re.findall(property_expression_pattern, form.text)
    
    if props_exprs:
        forms_exprs_props[form] = []
        
        for match in props_exprs:
            exprs = re.findall(expression_pattern, match[1])
            if exprs:
                for e in exprs:
                    forms_exprs_props[form].append(Expression(prop=match[0], condition_type=e[0], expression=e[1]))

In [14]:
logger.remove()
logger.add(sys.stdout, colorize=True, level="DEBUG", format="{message}")

for f in forms_exprs_props.keys():
    if forms_exprs_props[f]:
        print('-----------------    FORM   -------------------')
        for dependence in forms_exprs_props[f]:
            logger.opt(colors=True).debug('<blue>{prop}</blue> <red>|</red> <green>{expr}</green> from <cyan>{type}</cyan>', prop=dependence.prop, expr=dependence.expression, type=dependence.condition_type)

-----------------    FORM   -------------------
notificationTextDocMiss | !empty securityDataProvider.loggedInUser from renderedCondition
TypeApplicationRef | !empty securityDataProvider.loggedInUser from renderedCondition
TypeApplicationRef | true from readOnlyCondition
SignMiddleName | fields['SignFL'] eq true or fields['SignUL'] eq true from renderedCondition
sign_wiringDiagram_appl | view.viewId.contains('portal.xhtml') and ['UL','IP'].contains(fields.ApplicantType.Code) and !empty fields.wiringDiagram and fields.p_ContractTransferType.Code eq '7185643' and (!empty fields.id or (dataObjectController.instance.objectStatus.status.code eq 'ta03_Paused' and fields.SendDate&gt;=elUtils.toDate(2020,6,1))) from renderedCondition
sign_wiringDiagram_appl | view.viewId.contains('portal.xhtml') and ['UL','IP'].contains(fields.ApplicantType.Code) and !empty fields.wiringDiagram from requiredCondition
ScanRBPAct | fields['p_TypeTP'].Code eq '32880' and fields.SendDate&lt;elUtils.toDate(2020,6,1

contractGP | view.viewId.contains('portal.xhtml') and empty fields.contractGP and (dataObjectController.instance.objectStatus.status.code ne 'ta10_ContractExecuted' or dataObjectController.instance.objectStatus.status.code ne 'ta04_Canceled' or dataObjectController.instance.objectStatus.status.code ne 'ta11_ContractTerminated') from readOnlyCondition
ElectroCompany3 | fields.ORGNotFound ne true from renderedCondition
ElectroCompany3 | true from requiredCondition
contactphone | empty securityDataProvider.loggedInUser from requiredCondition
cBuildingaddrFact | fields.ApplicantType.value.fields.Code eq 'UL' ? 'Фактический адрес' : 'Адрес места жительства' from group
cBuildingaddrFact | !(empty securityDataProvider.loggedInUser or securityDataProvider.hasRole('tehprisEE_portalUserRegistrator')  or (securityDataProvider.hasRole('tehprisEE_ZayavkaTP')) and empty sessionContext.get('accession_user')) from readOnlyCondition
EPUPlanDoc | fields['p_TypeTP'].Code eq '32880' from renderedCondition

DeviceAddrNotFound | fields['p_DeviceAddrMatchWith'].Code eq '32886' from renderedCondition
PUSerialNumber | !empty securityDataProvider.loggedInUser from renderedCondition
sign_ScanDopDogovor | !empty securityDataProvider.loggedInUser from renderedCondition
sign_ScanDopDogovor | fields.p_ContractTransferType.Code eq '7185643' and !empty fields.ScanDopDogovor from requiredCondition
PaperGPDocsSendDate | fields.PaperGPDocsSend eq true from renderedCondition
PaperGPDocsSendDate | fields.PaperGPDocsSend eq true from readOnlyCondition
sign_ScanActTU | !empty securityDataProvider.loggedInUser from renderedCondition
sign_ScanActTU | fields.p_ContractTransferType.Code eq '7185643' and !empty fields.ScanActTU from requiredCondition
Filial | true from renderedCondition
Filial | true from requiredCondition
Filial | !(empty securityDataProvider.loggedInUser or (securityDataProvider.hasRole('tehprisEE_portalUserRegistrator')) or (securityDataProvider.hasRole('tehprisEE_ZayavkaTP')) and empty field

CorAccountGP | empty fields.CorAccountGP and isSimpleTPScheme eq true and fields.PlanDateProjectApplLoad &gt;=elUtils.today() and fields.ApplicantType.Code eq 'FL' and fields.SendDate&gt;=elUtils.toDate(2020,6,1) from requiredCondition
CorAccountGP | !empty fields.CorAccountGP and securityDataProvider.hasRole('tehprisEE_GpWorkP') or view.viewId.contains('portal.xhtml') or securityDataProvider.hasRole('tehprisEE_ZayavkaTP') from readOnlyCondition
SignFL | fields.ApplicantType.value.fields.Code eq 'FL' or fields.ApplicantType.value.fields.Code eq 'IP' from renderedCondition
CityaddrReg | fields.ApplicantType.value.fields.Code eq 'UL' ? 'Юридический адрес' : 'Адрес места регистрации' from group
CityaddrReg | true from renderedCondition
CityaddrReg | empty fields.BoroughaddrReg and empty fields.CityaddrReg from requiredCondition
CityaddrReg | !(empty securityDataProvider.loggedInUser or securityDataProvider.hasRole('tehprisEE_portalUserRegistrator') or (securityDataProvider.hasRole('tehpri

Regionaddr | !(securityDataProvider.hasRole('tehprisEE_portalUserRegistrator') or securityDataProvider.hasRole('tehprisEE_ZayavkaTP')) from renderedCondition
Regionaddr | empty securityDataProvider.loggedInUser or securityDataProvider.hasRole('tehprisEE_portalUserRegistrator') or (securityDataProvider.hasRole('tehprisEE_ZayavkaTP')) and empty fields.id from requiredCondition
Regionaddr | !(empty securityDataProvider.loggedInUser or securityDataProvider.hasRole('tehprisEE_portalUserRegistrator') or (securityDataProvider.hasRole('tehprisEE_ZayavkaTP')) and empty fields.id) from readOnlyCondition
SignPersData | (fields['SignFL'] eq true or ['UL'].contains(fields.ApplicantType.Code)) from renderedCondition
SignPersData | (fields['SignFL'] eq true or ['UL'].contains(fields.ApplicantType.Code)) and empty fields.id or fields.SendDate&gt;=elUtils.toDate(2020,6,1) from requiredCondition
IdAppealACCO | !empty securityDataProvider.loggedInUser from renderedCondition
FullPowerD |  (fields['p_TypeT

DULDoc | empty securityDataProvider.loggedInUser or securityDataProvider.hasRole('tehprisEE_portalUserRegistrator') or (securityDataProvider.hasRole('tehprisEE_ZayavkaTP')) and empty fields.id from requiredCondition
DULDoc | !(empty securityDataProvider.loggedInUser or (securityDataProvider.hasRole('tehprisEE_portalUserRegistrator')) or (securityDataProvider.hasRole('tehprisEE_ZayavkaTP')) and empty fields.id) from readOnlyCondition
RedresentMiddleName | fields.ApplicantType.value.fields.Code eq 'UL' from renderedCondition
AreaaddrFact | fields.ApplicantType.value.fields.Code eq 'UL' ? 'Фактический адрес' : 'Адрес места жительства' from group
AreaaddrFact | true from renderedCondition
AreaaddrFact | !(empty securityDataProvider.loggedInUser or securityDataProvider.hasRole('tehprisEE_portalUserRegistrator') or (securityDataProvider.hasRole('tehprisEE_ZayavkaTP')) and empty fields.id) from readOnlyCondition
ContractDateConfirm | !empty securityDataProvider.loggedInUser from renderedCondi

UninhabitedTP | fields['p_TypeTP'].Code eq '32880' from renderedCondition
ScanActTP_User | view.viewId.contains('portal.xhtml') and fields.ApplicantType.Code eq 'FL' and fields.p_ContractTransferType.Code eq '7185643' and !empty fields.ScanActTP_User and fields.SendDate&gt;=elUtils.toDate(2020,6,1) from signCondition
ScanActTP_User | !empty securityDataProvider.loggedInUser and isSimpleTPScheme ne true from renderedCondition
ScanActTP_User | securityDataProvider.hasRole('tehprisEE_User') and !empty fields.ScanActTP and isSimpleTPScheme ne true from requiredCondition
ScanActTP_User | !(dataObjectController.instance.objectStatus.status.code eq 'ta09_ActsSigning' and fields.p_ContractTransferType.Code eq '7185643' and !empty fields.ScanActTP and dashboardContext.context.vis_buttonSend or securityDataProvider.hasRole('tehprisEE_ZayavkaTP')) from readOnlyCondition
RedresentLastName | fields.ApplicantType.value.fields.Code eq 'UL' from renderedCondition
CorAccount | fields.ApplicantType.Code

RegionaddrFactRef | securityDataProvider.hasRole('tehprisEE_portalUserRegistrator')or (securityDataProvider.hasRole('tehprisEE_ZayavkaTP')) and empty fields.id from renderedCondition
RegionaddrFactRef | true from requiredCondition
RegionaddrFactRef | !(empty securityDataProvider.loggedInUser or securityDataProvider.hasRole('tehprisEE_portalUserRegistrator') or (securityDataProvider.hasRole('tehprisEE_ZayavkaTP')) and empty fields.id) from readOnlyCondition
Corpaddr | true from renderedCondition
Corpaddr | !(empty securityDataProvider.loggedInUser or securityDataProvider.hasRole('tehprisEE_portalUserRegistrator') or (securityDataProvider.hasRole('tehprisEE_ZayavkaTP')) and empty fields.id) from readOnlyCondition
GardenPersData | fields.GardenTP eq true and (view.viewId.contains('portal.xhtml') or (securityDataProvider.hasRole('tehprisEE_portalUserRegistrator'))) from renderedCondition
GardenPersData | fields.GardenTP eq true and view.viewId.contains('portal.xhtml') and empty fields.id o

GardenSprav | fields.GardenTP eq true and view.viewId.contains('portal.xhtml') and empty fields.id or fields.SendDate&gt;=elUtils.toDate(2020,6,1) from requiredCondition
PersDataAgree | true from requiredCondition
sign_TuDoc_appl | !empty securityDataProvider.loggedInUser and fields.isApplicantRefuse ne true and fields.p_ContractTransferType.Code eq '7185643' and ['UL','IP'].contains(fields.ApplicantType.Code) and !empty fields.TuDoc and isSimpleTPScheme ne true from renderedCondition
sign_TuDoc_appl | securityDataProvider.hasRole('tehprisEE_User') and (fields.ApplicantType.Code eq 'IP' or fields.ApplicantType.Code eq 'UL') and fields.p_ContractTransferType.Code eq '7185643' and !empty fields.TuDoc from requiredCondition
INNGP | fields.isSimpleTPScheme eq true and fields.ApplicantType.Code eq 'FL' from renderedCondition
INNGP | empty fields.INNGP and isSimpleTPScheme eq true and fields.PlanDateProjectApplLoad &gt;=elUtils.today() and fields.ApplicantType.Code eq 'FL' and fields.SendDat

CityaddrFact | true from renderedCondition
CityaddrFact | empty fields.BoroughaddrFact and empty fields.CityaddrFact from requiredCondition
CityaddrFact | !(empty securityDataProvider.loggedInUser or securityDataProvider.hasRole('tehprisEE_portalUserRegistrator') or (securityDataProvider.hasRole('tehprisEE_ZayavkaTP')) and empty fields.id) from readOnlyCondition
firstName | empty securityDataProvider.loggedInUser or securityDataProvider.hasRole('tehprisEE_portalUserRegistrator') or (securityDataProvider.hasRole('tehprisEE_ZayavkaTP')) and empty fields.id from requiredCondition
DeviceAddrRayon | fields['p_DeviceAddrMatchWith'].Code eq '32886' and fields.DeviceAddrNotFound eq true from renderedCondition
ogrnGP | fields.isSimpleTPScheme eq true and fields.ApplicantType.Code eq 'FL' from renderedCondition
ogrnGP | empty fields.ogrnGP and isSimpleTPScheme eq true and fields.PlanDateProjectApplLoad &gt;=elUtils.today() and fields.ApplicantType.Code eq 'FL' and fields.SendDate&gt;=elUtils.toD

ApplAccountGP | !empty fields.ApplAccountGP and securityDataProvider.hasRole('tehprisEE_GpWorkP') or view.viewId.contains('portal.xhtml') or securityDataProvider.hasRole('tehprisEE_ZayavkaTP') from readOnlyCondition
lowVoltageTP | !empty securityDataProvider.loggedInUser from renderedCondition
AgreeUsePublicPropertyDoc | fields['p_TypeTP'].Code eq '32880' and fields['GardenTP'] eq true and !empty fields.id and fields.SendDate&lt;elUtils.toDate(2020,6,1) from renderedCondition
StagesTP | fields['p_TypeTP'].Code eq '32880' from renderedCondition
StagesTP | fields['p_TypeTP'].Code eq '32880' from requiredCondition
p_ContractType | fields['p_TypeTP'].Code eq '32880' or fields['p_TypeTP'].Code eq '32896' from requiredCondition
p_PaymentType | fields['p_TypeTP'].Code eq '32880' and (['UL','IP'].contains(fields.ApplicantType.Code)) and (fields['FullPowerD'] &gt;= 15 and fields['FullPowerD'] &lt;= 150) or ((empty fields.id or fields.SendDate &gt;= elUtils.toDate(2020,6,1))) and fields['FullPow

TUApplicantComment | fields._Status eq 'ta07_TUImplementationCheck' and !empty fields.TUCompanyCommentDoc from renderedCondition
TUApplicantComment | securityDataProvider.hasRole('tehprisEE_User') and dataObjectController.instance.objectStatus.status.code eq 'ta07_TUImplementationCheck' and !empty fields.TUCompanyCommentDoc and dashboardContext.context.vis_buttonSend from requiredCondition
TUApplicantComment | !empty fields.ScanActTU and !fields.TUCompanyComment.isEmptyAwareDeleted() and dashboardContext.context.vis_buttonSend from readOnlyCondition
TPInstruction | fields.isSimpleTPScheme eq true from renderedCondition
TPInstruction | securityDataProvider.hasRole('tehprisEE_ZayavkaTP') and fields.isSimpleTPScheme eq true and fields._Status eq 'ta02_Accepted' from requiredCondition
RegistrIPDoc | fields.ApplicantType.value.fields.Code eq 'IP' from renderedCondition
RegistrIPDoc | fields.ApplicantType.value.fields.Code eq 'IP' from requiredCondition
HeadFirstName | fields.ApplicantType.v

Houseaddr | true from renderedCondition
Houseaddr | empty securityDataProvider.loggedInUser or securityDataProvider.hasRole('tehprisEE_portalUserRegistrator') or (securityDataProvider.hasRole('tehprisEE_ZayavkaTP')) and empty fields.id from requiredCondition
Houseaddr | !(empty securityDataProvider.loggedInUser or securityDataProvider.hasRole('tehprisEE_portalUserRegistrator') or (securityDataProvider.hasRole('tehprisEE_ZayavkaTP')) and empty fields.id) from readOnlyCondition
IdAppealDZO | !empty securityDataProvider.loggedInUser from renderedCondition
ToleranceDate | !empty securityDataProvider.loggedInUser from renderedCondition
wiringDiagram | (fields.ApplicantType.value.fields.Code eq 'FL' and view.viewId.contains('portal.xhtml')) or (securityDataProvider.hasRole('tehprisEE_portalUserRegistrator')) and fields.p_ContractTransferType.Code eq '7185643' and !empty fields.wiringDiagram and fields._Status eq 'ta07_TUImplementationCheck' and fields.SendDate&gt;=elUtils.toDate(2020,6,1) fr

ScanNotificationLetterSO | !empty securityDataProvider.loggedInUser from renderedCondition
GardenPersData | fields.GardenTP eq true and (view.viewId.contains('portal.xhtml') or (securityDataProvider.hasRole('tehprisEE_portalUserRegistrator'))) from renderedCondition
GardenPersData | fields.GardenTP eq true and view.viewId.contains('portal.xhtml') and empty fields.id or fields.SendDate&gt;=elUtils.toDate(2020,6,1) from requiredCondition
SingleLineSchemeDoc | ['56991','56992','56993','56994'].contains(fields.p_FullVoltage.Code) from renderedCondition
SingleLineSchemeDoc | ['56991','56992','56993','56994'].contains(fields.p_FullVoltage.Code) and empty fields.id or fields.SendDate&gt;=elUtils.toDate(2020,6,1) from requiredCondition
ScanDopDogovorUser | view.viewId.contains('portal.xhtml') and fields.ApplicantType.Code eq 'FL' and fields.p_ContractTransferType.Code eq '7185643' and !empty fields.ScanDopDogovorUser and fields.SendDate&gt;=elUtils.toDate(2020,6,1) from signCondition
ScanDopDo

AreaaddrReg | true from renderedCondition
AreaaddrReg | !(empty securityDataProvider.loggedInUser or securityDataProvider.hasRole('tehprisEE_portalUserRegistrator') or (securityDataProvider.hasRole('tehprisEE_ZayavkaTP')) and empty fields.id) from readOnlyCondition
invoiceTP | !empty securityDataProvider.loggedInUser from renderedCondition
invoiceTP | securityDataProvider.hasRole('tehprisEE_ZayavkaTP') and fields.isSimpleTPScheme eq true and dataObjectController.instance.objectStatus.status.code eq 'ta02_Accepted' from requiredCondition
invoiceTP | !securityDataProvider.hasRole('tehprisEE_ZayavkaTP') from readOnlyCondition
DeviceAddrLocality | fields['p_DeviceAddrMatchWith'].Code eq '32886' and fields.DeviceAddrNotFound eq true from renderedCondition
DeviceAddrLocality | !empty fields.DeviceAddrRayon and empty fields.DeviceAddrCity from requiredCondition
sign_ScanNotificationLetterSO_SO | !empty securityDataProvider.loggedInUser from renderedCondition
sign_ScanNotificationLetterSO_SO |

p_RejoinsVoltage | fields['p_TypeTP'].Code eq '32880' or fields['p_TypeTP'].Code eq '32896' from renderedCondition
p_RejoinsVoltage | fields['p_TypeTP'].Code eq '32880' or fields['p_TypeTP'].Code eq '32896' from requiredCondition
ActualConnect | fields.SendDate&gt;=elUtils.toDate(2020,6,15) and (securityDataProvider.hasRole('tehprisEE_ZayavkaTP')) and (fields._Status eq 'ta09_ActsSigning') from requiredCondition
ApplicationDoc | true from renderedCondition
DZOInfoLabel | view.viewId.contains('portal.xhtml') from renderedCondition
isSimpleTPScheme | !view.viewId.contains('portal.xhtml') from renderedCondition
isSimpleTPScheme | true from readOnlyCondition
isSimpleTPScheme | !empty fields.isSimpleTPScheme ? fields.isSimpleTPScheme : ((empty fields.id or fields.SendDate&gt;=elUtils.toDate(2020,6,1)) and fields['p_TypeTP'].Code eq '32880' and ((fields.ApplicantType.value.fields.Code eq 'FL' and fields.p_TypeOfLoad.value.fields.Code eq '56997' and ['32837'].contains(fields.p_ReliabilityCate

Areaaddr | !(empty securityDataProvider.loggedInUser or securityDataProvider.hasRole('tehprisEE_portalUserRegistrator') or (securityDataProvider.hasRole('tehprisEE_ZayavkaTP')) and empty fields.id) from readOnlyCondition
ScanActBooking_appl | !empty securityDataProvider.loggedInUser from renderedCondition
ScanActBooking_appl | securityDataProvider.hasRole('tehprisEE_User') and !empty fields.ScanActBooking from requiredCondition
Boroughaddr | true from renderedCondition
Boroughaddr | empty fields.Boroughaddr and empty fields.Cityaddr from requiredCondition
Boroughaddr | !(empty securityDataProvider.loggedInUser or securityDataProvider.hasRole('tehprisEE_portalUserRegistrator') or (securityDataProvider.hasRole('tehprisEE_ZayavkaTP')) and empty fields.id) from readOnlyCondition
DateProjectGPLoad | true from readOnlyCondition
emailAddress | empty securityDataProvider.loggedInUser from requiredCondition
emailAddress | !empty fields.emailAddress or !empty fields.id ? fields.emailAddress : ( 

ogrnGP | empty fields.ogrnGP and isSimpleTPScheme eq true and fields.PlanDateProjectApplLoad &gt;=elUtils.today() and fields.ApplicantType.Code eq 'FL' and fields.SendDate&gt;=elUtils.toDate(2020,6,1) from requiredCondition
ogrnGP | !empty fields.ogrnGP from readOnlyCondition
sign_ScanActTP | !empty securityDataProvider.loggedInUser from renderedCondition
sign_ScanActTP | fields.p_ContractTransferType.Code eq '7185643' and !empty fields.ScanActTP from requiredCondition
SignFirstName | fields['SignFL'] eq true or fields['SignUL'] eq true from renderedCondition
SignFirstName | fields['SignFL'] eq true or fields['SignUL'] eq true from requiredCondition
Region | true from renderedCondition
Region | true from requiredCondition
Region | !(empty securityDataProvider.loggedInUser or (securityDataProvider.hasRole('tehprisEE_portalUserRegistrator')) or (securityDataProvider.hasRole('tehprisEE_ZayavkaTP')) and empty fields.id) from readOnlyCondition
ReasonCancelContract | (fields.AddAgreementReas

Indexaddr | !(empty securityDataProvider.loggedInUser or securityDataProvider.hasRole('tehprisEE_portalUserRegistrator') or (securityDataProvider.hasRole('tehprisEE_ZayavkaTP')) and empty fields.id) from readOnlyCondition
ScanDopDogovor | !empty securityDataProvider.loggedInUser from renderedCondition
p_ApplicationSendType | !empty securityDataProvider.loggedInUser from renderedCondition
CorAccountGP | fields.isSimpleTPScheme eq true and fields.ApplicantType.Code eq 'FL' from renderedCondition
CorAccountGP | empty fields.CorAccountGP and isSimpleTPScheme eq true and fields.PlanDateProjectApplLoad &gt;=elUtils.today() and fields.ApplicantType.Code eq 'FL' and fields.SendDate&gt;=elUtils.toDate(2020,6,1) from requiredCondition
CorAccountGP | !empty fields.CorAccountGP and securityDataProvider.hasRole('tehprisEE_GpWorkP') or view.viewId.contains('portal.xhtml') or securityDataProvider.hasRole('tehprisEE_ZayavkaTP') from readOnlyCondition
RedresentDocNum | fields.ApplicantType.value.fields

DULnum | fields.ApplicantType.Code eq 'FL' or fields.ApplicantType.Code eq 'IP' from renderedCondition
DULnum | empty securityDataProvider.loggedInUser or securityDataProvider.hasRole('tehprisEE_portalUserRegistrator') or (securityDataProvider.hasRole('tehprisEE_ZayavkaTP')) and empty fields.id from requiredCondition
DULnum | !(empty securityDataProvider.loggedInUser or securityDataProvider.hasRole('tehprisEE_portalUserRegistrator') or (securityDataProvider.hasRole('tehprisEE_ZayavkaTP')) and empty fields.id) from readOnlyCondition
p_DeviceAddrMatchWith | true from requiredCondition
p_DeviceAddrMatchWith | empty fields.p_DeviceAddrMatchWith ? ( fields.ApplicantType.Code eq 'FL' or fields.ApplicantType.Code eq 'IP' ? '32924' : (fields.ApplicantType.Code eq 'UL' ? '32926' :  fields.p_DeviceAddrMatchWith) ) : fields.p_DeviceAddrMatchWith from defaultValue
notificationDateDocMiss | !empty securityDataProvider.loggedInUser from renderedCondition
ApplAccountGP | fields.isSimpleTPScheme eq tr

RejoinsPowerD | fields['p_TypeTP'].Code eq '32880' or fields['p_TypeTP'].Code eq '32896' from requiredCondition
IdAppealACCO | !empty securityDataProvider.loggedInUser from renderedCondition
FullPowerD |  (fields['p_TypeTP'].Code eq '32880' or fields['p_TypeTP'].Code eq '32897') from renderedCondition
FullPowerD | fields['p_TypeTP'].Code eq '32880' or fields['p_TypeTP'].Code eq '32897' from requiredCondition
AccountGP | fields.isSimpleTPScheme eq true and fields.ApplicantType.Code eq 'FL' from renderedCondition
AccountGP | empty fields.AccountGP and isSimpleTPScheme eq true and fields.PlanDateProjectApplLoad &gt;=elUtils.today() and fields.ApplicantType.Code eq 'FL' and fields.SendDate&gt;=elUtils.toDate(2020,6,1) from requiredCondition
AccountGP | !empty fields.AccountGP and securityDataProvider.hasRole('tehprisEE_GpWorkP') or view.viewId.contains('portal.xhtml') or securityDataProvider.hasRole('tehprisEE_ZayavkaTP') from readOnlyCondition
DateCancelContract | (fields.AddAgreementReas

p_EarlierVoltage | fields['p_TypeTP'].Code eq '32880' and fields['p_ReasonTP'].Code eq '956' from renderedCondition
p_EarlierVoltage | not empty fields['EarlierPowerD'] and fields['EarlierPowerD'] ne 0 from requiredCondition
CorpaddrFact | fields.ApplicantType.value.fields.Code eq 'UL' ? 'Фактический адрес' : 'Адрес места жительства' from group
CorpaddrFact | true from renderedCondition
CorpaddrFact | !(empty securityDataProvider.loggedInUser or securityDataProvider.hasRole('tehprisEE_portalUserRegistrator') or (securityDataProvider.hasRole('tehprisEE_ZayavkaTP')) and empty fields.id) from readOnlyCondition
StatementNumber | !empty securityDataProvider.loggedInUser from renderedCondition
ogrnIP | fields.ApplicantType.Code eq 'IP' from renderedCondition
RedresentFirstName | fields.ApplicantType.value.fields.Code eq 'UL' from renderedCondition
GardenTP | fields['p_TypeTP'].Code eq '32880' and fields.ApplicantType.value.fields.Code eq 'UL' from renderedCondition
ContractDatePerfomance | !

designDoc | (fields.ApplicantType.value.fields.Code eq 'FL' and view.viewId.contains('portal.xhtml')) or (securityDataProvider.hasRole('tehprisEE_portalUserRegistrator')) and fields.p_ContractTransferType.Code eq '7185643' and !empty fields.designDoc and fields._Status eq 'ta07_TUImplementationCheck' and fields.SendDate&gt;=elUtils.toDate(2020,6,1) from signCondition
designDoc | !empty securityDataProvider.loggedInUser from renderedCondition
TUApplicantComment | fields._Status eq 'ta07_TUImplementationCheck' and !empty fields.TUCompanyCommentDoc from renderedCondition
TUApplicantComment | securityDataProvider.hasRole('tehprisEE_User') and dataObjectController.instance.objectStatus.status.code eq 'ta07_TUImplementationCheck' and !empty fields.TUCompanyCommentDoc and dashboardContext.context.vis_buttonSend from requiredCondition
TUApplicantComment | !empty fields.ScanActTU and !fields.TUCompanyComment.isEmptyAwareDeleted() and dashboardContext.context.vis_buttonSend from readOnlyConditio

DateTUCorrectApplicant | !empty fields.ScanActTU and dashboardContext.context.vis_buttonSend from readOnlyCondition
p_assoTPPortalProccessResult | !empty securityDataProvider.loggedInUser from renderedCondition
UninhabitedTP | fields['p_TypeTP'].Code eq '32880' from renderedCondition
isApplRefuseActTP | true or fields.Login eq securityDataProvider.loggedInUser.loginName and fields._Status eq 'ta09_ActsSigning' and ((!empty fields.ScanActTP and empty fields.ScanActTP_User)) from renderedCondition
isApplRefuseActTP | (false and( fields.Login ne securityDataProvider.loggedInUser.loginName or fields._Status ne 'ScanActTP_User' )) or  (dataObjectController.instance.objectStatus.status.code eq 'ta10_ContractExecuted'  or dataObjectController.instance.objectStatus.status.code eq 'ta04_Canceled'  or dataObjectController.instance.objectStatus.status.code eq 'ta11_ContractTerminated' ) from readOnlyCondition
ScanActTP_User | view.viewId.contains('portal.xhtml') and fields.ApplicantType.Code eq '

Houseaddr | true from renderedCondition
Houseaddr | empty securityDataProvider.loggedInUser or securityDataProvider.hasRole('tehprisEE_portalUserRegistrator')  or (securityDataProvider.hasRole('tehprisEE_ZayavkaTP')) and empty fields.id from requiredCondition
Houseaddr | !(empty securityDataProvider.loggedInUser or securityDataProvider.hasRole('tehprisEE_portalUserRegistrator')and empty fields.id) from readOnlyCondition
DeviceAddrRegionRef | fields['p_DeviceAddrMatchWith'].Code eq '32886' and fields.DeviceAddrNotFound eq true from renderedCondition
DeviceAddrRegionRef | fields['p_DeviceAddrMatchWith'].Code eq '32886' from requiredCondition
IdAppealDZO | !empty securityDataProvider.loggedInUser from renderedCondition
EVLabel | fields.isSimpleTPScheme eq true and view.viewId.contains('portal.xhtml') and (empty fields.id or (dataObjectController.instance.objectStatus.status.code eq 'ta03_Paused')) from renderedCondition
ToleranceDate | !empty securityDataProvider.loggedInUser from rendere

In [8]:


for f in forms_expressions.keys():
    if forms_expressions[f]:
        print('-----------------    FORM   -------------------')
        for exp in forms_expressions[f]:
            if exp.tag == 'renderedCondition':
                logger.opt(colors=True).debug('Expression <green>{expr}</green> from <red>{form}</red>', expr=exp.value, form=exp.tag)

NameError: name 'forms_expressions' is not defined

In [ ]:
for f in forms_expressions.keys():
    if forms_expressions[f]:
        print('-----------------    FORM   -------------------')
        for exp in forms_expressions[f]:
            if exp.tag == 'requiredCondition':
                logger.opt(colors=True).debug('Expression <green>{expr}</green> from <red>{form}</red>', expr=exp.value, form=exp.tag)

In [ ]:
for f in forms_expressions.keys():
    if forms_expressions[f]:
        print('-----------------    FORM   -------------------')
        for exp in forms_expressions[f]:
            if exp.tag == 'readOnlyCondition':
                logger.opt(colors=True).debug('Expression <green>{expr}</green> from <red>{form}</red>', expr=exp.value, form=exp.tag)

In [ ]:
for f in forms_expressions.keys():
    if forms_expressions[f]:
        print('-----------------    FORM   -------------------')
        for exp in forms_expressions[f]:
            if exp.tag == 'defaultValue':
                logger.opt(colors=True).debug('Expression <green>{expr}</green> from <red>{form}</red>', expr=exp.value, form=exp.tag)

In [ ]:
import networkx as nx
from matplotlib import pyplot as plt

In [ ]:
g = nx.Graph()

In [ ]:
g.add_edge('1', '2')

In [ ]:
nx.draw_networkx(g)

In [ ]:
re.findall(r'.+\.(\w+)' , '../data_samples/Заявка_на_установку_замену_ПУ_522_прод_01_03_2021.xml')[0]

In [ ]:
from enum import Enum

scope_open_re = re.compile(r'^\(')
scope_close_re = re.compile(r'^\)')
value_re = re.compile(r'^[\w.\']*(?:\(\'\w*\'\))?')
not_re = re.compile(r'^!')
and_re = re.compile(r'^and')
or_re = re.compile(r'^or')
empty_re = re.compile(r'^empty')


class JavaELTokenType(Enum):
    SCOPE_OPEN = 1
    SCOPE_CLOSE = 2
    LVALUE = 3
    RVALUE = 4
    NOT = 5
    AND = 6
    OR = 7
    EMPTY = 8


JavaELToken = namedtuple('JavaELToken', ('token_type', 'token_value'))


def tokenize_expression(expression: str):
    """
    Parse JavaEL expression by tokens
    :param expression:
    :return: List[JavaELToken]
    """

    scope_open_match = None
    scope_close_match = None
    value_match = None
    not_match = None
    or_match = None
    and_match = None
    empty_match = None

    tokenized = []
    offset = 0
    
    while len(expression):
        
        scope_open_match = scope_open_re.findall(expression)
        scope_close_match = scope_close_re.findall(expression)
        value_match = value_re.findall(expression)
        not_match = not_re.findall(expression)
        or_match = or_re.findall(expression)
        and_match = and_re.findall(expression)
        empty_match = empty_re.findall(expression)
        
        if scope_open_match and scope_open_match[0]:
            tokenized.append(JavaELToken(token_type=JavaELTokenType.SCOPE_OPEN, token_value='('))
            offset = len(scope_open_match[0])
        elif scope_close_match and scope_close_match[0]:
            tokenized.append(JavaELToken(token_type=JavaELTokenType.SCOPE_CLOSE, token_value=')'))
        elif not_match and not_match[0]:
            tokenized.append(JavaELToken(token_type=JavaELTokenType.NOT, token_value='!'))
            offset = len(not_match[0])
        elif or_match and or_match[0]:
            tokenized.append(JavaELToken(token_type=JavaELTokenType.OR, token_value='or'))
            offset = len(or_match[0])
        elif and_match and and_match[0]:
            tokenized.append(JavaELToken(token_type=JavaELTokenType.AND, token_value='and'))
            offset = len(and_match[0])
        elif empty_match and empty_match[0]:
            tokenized.append(JavaELToken(token_type=JavaELTokenType.EMPTY, token_value='empty'))
            offset = len(empty_match[0])
        elif value_match and value_match[0]:
            if '\'' in value_match[0] or value_match[0] == 'true' or value_match[0] == 'false':
                tokenized.append(JavaELToken(token_type=JavaELTokenType.RVALUE, token_value=value_match[0]))
            else:
                tokenized.append(JavaELToken(token_type=JavaELTokenType.LVALUE, token_value=value_match[0]))
            offset = len(value_match[0])
        
        expression = expression[offset:].strip()
        
        scope_open_match = None
        scope_close_match = None
        value_match = None
        not_match = None
        or_match = None
        and_match = None
        empty_match = None
        
        logger.opt(colors=True).debug('<red>{exp}</red> | <green>{ofst}</green> | <blue>{l}</blue>', exp = expression, ofst=offset, l=len(expression))
    return tokenized

In [ ]:
tokenize_expression('!empty fields.TypeApplicationRef ? fields.TypeApplicationRef : \'tehprisEE_Install_Replace_PU522\'')

In [ ]:
not_re.findall("!(empty securityDataProvider.loggedInUser or securityDataProvider.hasRole('tehprisEE_portalUserRegistrator')and empty fields.id)")

In [ ]:
value_re.findall("securityDataProvider.loggedInUser or securityDataProvider.hasRole('tehprisEE_portalUserRegistrator')and empty fields.id)")

In [ ]:
import pandas as pd

In [ ]:
df = pd.DataFrame(columns=['col1', 'col2', 'col3'])

In [ ]:
df

In [ ]:
r = re.findall(r'^(\w+)\.', "true")

In [ ]:
r[0]

In [9]:
re.search(r'objectForm name=\"(.+?)\"', "<classAccessRules/><forms>&lt;?xml version="1.0" encoding="UTF-8" standalone="yes"?>
&lt;objectForm name="Реестр гарантирующих поставщиков" sectionStyle="tab">
    &lt;otherForms/>
    &lt;fields>
        &lt;entry>
            &lt;key>sign_SalesContractIP_GP&lt;/key>
            &lt;value xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:type="fileFormField" fileTypes="sig" fileTypesErrorMessage="Недопустимый формат файла. Допустимые форматы: sig." fileMaxSize="5" fileMaxSizeErrorMessage="Размер файла превышает допустимый размер - 5 Мб." simple="true" noVersionRestore="true" code="sign_SalesContractIP_GP" name="Файл УКЭП для &amp;quot;Проект договора купли-продажи электроэнергии и оказания услуг по передаче электроэнергии с индивидуальным предпринимателем&amp;quot;" section="Общие" sectionOrder="1" index="39" group="Проекты договоров" col="1" row="28" labelWidth="420" labelStyle="text-align: right; " resetHidden="false" viewType="standard">
                &lt;facesValidatorParams/>
                &lt;resetFields/>")

SyntaxError: invalid syntax (649475129.py, line 1)

dashboardContext.context.vis_buttonSend == (empty fields.id or !securityDataProvider.hasRole('tehprisEE_User')or(dataObjectController.instance.objectStatus.status.code eq 'ta03_Paused')or(dataObjectController.instance.objectStatus.status.code eq 'ta05_ContractOnReviewApplicant' and fields.p_ContractTransferType.Code eq '7185643' and fields.isApplicantRefuse ne true and (!empty fields['ScanDogovor'] and empty fields['ScanDogovorCL'] or !empty fields['ScanDopDogovor'] and empty fields['ScanDopDogovorUser']))or(dataObjectController.instance.objectStatus.status.code eq 'ta06_ContractSigned' and fields.p_ContractTransferType.Code eq '7185643' and !empty fields['DateCompTechnical_Company'] and empty fields['DateCompTechnical_applicant'])or(dataObjectController.instance.objectStatus.status.code eq 'ta07_TUImplementationCheck' and fields.p_ContractTransferType.Code eq '7185643' and ((!empty fields['ScanActTU'] and empty fields['ScanActTU_User']) or (empty fields['ScanActTU'] and !empty fields['TUCompanyCommentDoc'] and empty fields['DateTUCorrectApplicant'])))or(dataObjectController.instance.objectStatus.status.code eq 'ta09_ActsSigning' and fields.p_ContractTransferType.Code eq '7185643' and !empty fields['ScanActTP'] and empty fields['ScanActTP_User']))

!(empty securityDataProvider.loggedInUser or securityDataProvider.hasRole('tehprisEE_portalUserRegistrator') or (securityDataProvider.hasRole('tehprisEE_ZayavkaTP')) and empty sessionContext.get('accession_user'))

empty securityDataProvider.loggedInUser or securityDataProvider.hasRole('tehprisEE_portalUserRegistrator') or (securityDataProvider.hasRole('tehprisEE_ZayavkaTP')) and empty fields.id

In [1]:
class C:
    a = 1
    b = 2

In [2]:
c = C()

In [3]:
c_id = id(c)

In [4]:
c_id

4425770896

In [5]:
import ctypes

In [6]:
ctypes.cast(c_id, ctypes.py_object).value